### PR TITLE
[options] Add authentication option `netrc-cmd`

### LINKED ISSUE #1706: Use authinfo.gpg instead of netrc for authentication
### Checklist

- [X] I'm reporting a feature request
- [X] I've verified that I'm running yt-dlp version **2021.11.10.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

Currently, there doesn't seem to be a way to use a gpg-encrypted `~/.authinfo.gpg` file instead of an unencrypted `~/.netrc` file. I use an `authinfo.gpg` file as it's a common way of providing credentials in Emacs. Encryption is obviously preferable over plain-text files for account credentials. I understand that including GPG encryption directly could be considered feature creep, so I'd suggest providing a flag provides the credentials via command output, similar to how `isync` does it. Here's an example of how I use this feature in `isync`'s config:

```
PassCmd "gpg --decrypt ~/.authinfo.gpg 2> /dev/null | grep outlook.office365.com | grep domain.com | sed 's/.*password \\([^ ]\\+\\).*/\\1/'"
```

Perhaps the netrc file could be provided like so:

```
--netrc-cmd "gpg --decrypt ~/.authinfo.gpg 2> /dev/null"
```

The rest of the netrc functionality would work as usual: the `netrc-cmd` flag would expect a decrypted netrc file as the return value.

When trying to use the `~/authinfo.gpg` file directly as-is, a predictable error is shown as the file cannot be parsed:

```
[debug] Command-line config: ['-Uv', 'https://twitch.tv/redacted']
[debug] User config file: /home/john/.config/yt-dlp/config
[debug] User config: ['--netrc', '--netrc-location', '~/.authinfo.gpg']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, err utf-8, pref UTF-8
[debug] yt-dlp version 2021.11.10.1
[debug] Python version 3.9.7 (CPython 64bit) - Linux-5.15.2-arch1-1-x86_64-with-glibc2.33
[debug] exe versions: ffmpeg 4.4.1 (setts), ffprobe 4.4.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
yt-dlp is up to date (2021.11.10.1)
ERROR: 'utf-8' codec can't decode byte 0x85 in position 0: invalid start byte
Traceback (most recent call last):
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1311, in wrapper
    return func(self, *args, **kwargs)
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1336, in __extract_info
    ie_result = ie.extract(url)
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 593, in extract
    self.initialize()
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 501, in initialize
    self._real_initialize()
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/extractor/twitch.py", line 58, in _real_initialize
    self._login()
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/extractor/twitch.py", line 61, in _login
    username, password = self._get_login_info()
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 1219, in _get_login_info
    username, password = self._get_netrc_login_info(netrc_machine)
  File "/home/john/.local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 1191, in _get_netrc_login_info
    info = netrc.netrc(file=netrc_file).authenticators(netrc_machine)
  File "/usr/lib/python3.9/netrc.py", line 30, in __init__
    self._parse(file, fp, default_netrc)
  File "/usr/lib/python3.9/netrc.py", line 39, in _parse
    toplevel = tt = lexer.get_token()
  File "/usr/lib/python3.9/shlex.py", line 109, in get_token
    raw = self.read_token()
  File "/usr/lib/python3.9/shlex.py", line 140, in read_token
    nextchar = self.instream.read(1)
  File "/usr/lib/python3.9/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 0: invalid start byte
```

### PR DESCRIPTION
**Pull request by NDagestad**
This adds a new option to specify a command that will be ran by yt-dlp to get the username and password that an extractor needs.
For now I kept the logic of having the passord on the first line of the output, and the username on the second one, I'll change it if you have a specific format in mind you prefer. 
I also updated the README, is there another place where documentation should be updated ?

As requested in the issue I opened, I also added the exclusion between the various credential related options.
The exclusion mechanism seems a bit clunky to me, but I can't think of another way that wouldn't be a lot of different comparisons. Maybe there is some python magic I don't know about that makes it possible to do an exclusive comparaison of 3 boolean variables 🤔

Fixes #1706

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### Summary
🆕📝🛠️

This pull request adds a new feature to `yt-dlp` that allows users to provide credentials for an extractor using a custom shell command. It introduces a new option `--netrc-cmd` and updates the relevant code and documentation to support this feature.

> _If you need to log in to extract_
> _But don't want to store your login facts_
> _You can use `--netrc-cmd`_
> _To run a shell command_
> _And pass your credentials in tact_

### Walkthrough
*  Add and document a new option `--netrc-cmd` to allow the user to provide credentials for an extractor using a custom shell command ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R53), [link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R908-R909), [link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R1223-R1230), [link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-c2d761a809b2c0014b1f78cff8163e4fe99be9eb6b779663f8a521990b35688cR701-R704))
*  Validate that the user does not specify more than one of the mutually exclusive options for providing credentials: `--username` and `--password`, `--netrc-cmd`, or `--netrc` in `yt_dlp/__init__.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-368601fd75cc1f1f9a4427938fe6381ffbab2ed89434cabb25d9346b43f037b0L190-R195))
*  Pass the value of the `--netrc-cmd` option to the extractor classes in `yt_dlp/__init__.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-368601fd75cc1f1f9a4427938fe6381ffbab2ed89434cabb25d9346b43f037b0R737))
*  Import and use the `subprocess` module to execute the shell command defined by the user in `yt_dlp/extractor/common.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6R16), [link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6R38))
*  Update the password hint message to include the `--netrc-cmd` option in `yt_dlp/extractor/common.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6L528-R530))
*  Implement a new method `_get_netrc_cmd_info` to parse the output of the shell command and return the username and password in `yt_dlp/extractor/common.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6R1308-R1326))
*  Modify the logic of the `_get_login_info` method to try the `--netrc-cmd` option before the `--netrc` option if the user did not provide the username and password manually in `yt_dlp/extractor/common.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/6682/files?diff=unified&w=0#diff-b7f6633815316063666fc3ecebb970e68fdcb5ee01beba03d04c01830e55e8a6L1311-R1344))

### CODE REVIEW

**Comment by pukkandan** [comment]
For ref: some of the disucussion is in https://github.com/yt-dlp/yt-dlp/issues/6676

**Comment by pukkandan** [comment]
> I also updated the README, is there another place where documentation should be updated ?

https://github.com/yt-dlp/yt-dlp/blob/68be95bd0ca3f76aa63c9812935bd826b3a42e53/yt_dlp/YoutubeDL.py#L192


> The exclusion mechanism seems a bit clunky to me, but I can't think of another way that wouldn't be a lot of different comparisons. Maybe there is some python magic I don't know about that makes it possible to do an exclusive comparaison of 3 boolean variables 🤔

```py
sum(map(bool, (opts.usenetrc, opts.netrc_cmd, opts.username))) > 1
```

> For now I kept the logic of having the passord on the first line of the output, and the username on the second one, I'll change it if you have a specific format in mind you prefer.

See https://github.com/yt-dlp/yt-dlp/issues/6676#issuecomment-1491427891, but I am open to discussion

**Comment by NDagestad** [comment]
From #6676 :

> I was thinking we expect netrc format? So `--netrc-cmd "cat ~/.netrc"` will work and other commands can return multiple machines if needed without parsing extractor key.

I though about doing that in the beginning, but it was more work and it mean every command would need to format their output like the netrc command. I can see the usefulness of being able to have the command be a nearly drop in replacement for the encrypted .netrc file. But the netrc library takes a path as argument, and that is surprisingly annoying to work around.
I have a  few solutions that can work, but they all have their drawbacks:
* Use a plain file instead of the FIFO. This has the disadvantage of writing the credentials to the disk, which I would imagine most people who don't want their password in a clear text file won't be very happy about.
* Create a FIFO, redirect the output to it and pass it to the netrc lib. Nothing is written on disk but for this, we need to use the low level os.{open,read,write} functions, which I think are not portable (it looks like we want to support Windows as well and not just unixes).
* Parse the netrc format by hand <- not a big fan of this idea even if it looks like a very simple format
* Patch the netrc library to be able to take a file object instead of a path. This would be the best of these options, but it might take a very long time to trickle down in the different distributions if it is accepted upstream at all.

In the end, I think it is better to have a very simple format for the command output, and rather have people do the parsing in the command itself, something like `gpg -d ~/.authinfo.gpg | sed -n 's/machine {} login \(.*\) password \(.*\)/\2\n\1/p'` could be enough for that kind of usecase.
But the format "password\nusername" is not something I strongly attached to, 

> `%(...)s` support can also be added in future if needed like we did for `--exec`. `{}` is fine for now. But should it be replaced with `ie_key()` or with `_NETRC_MACHINE`? I think the latter is better.

the `_get_netrc_login_info` handler function uses `_NETRC_MACHINE`, I think it makes sens for them to both uses the same selector, the  `--netrc-cmd "cat ~/.netrc"` behaviour would not work for all extractors if they uses different ones. (e.g. Nebula at least has different values for it)

**Comment by pukkandan** [comment]
> But the netrc library takes a path as argument, and that is surprisingly annoying to work around.

Untested, but something like this should work:

```py
class stream_netrc(netrc.netrc):
    def __init__(self, stream):
        self.hosts = {}
        self.macros = {}
        self._parse('-', stream, False)
```

Ref: https://github.com/python/cpython/blob/3.11/Lib/netrc.py#L67-L78


> I think it makes sense for them to both uses the same selector,

Agreed

**Comment by NDagestad** [comment]
> Untested, but something like this should work:
>
> ...
> 
> Ref: https://github.com/python/cpython/blob/3.11/Lib/netrc.py#L67-L78

Oh well, that worked perfectly! 
Do you want that class somewhere accessible so it can be reused by other parts of the yt-dlp ?

**Comment by pukkandan** [comment]
put it in `utils.py`

**Comment by NDagestad** [comment]
Sorry I didn't notify that I applied the changes suggested, so I don't know I you have seen it.
No rush though.

**Comment by NDagestad** [comment]
Rebased on latest commits; is there something blocking for this to get merged ?

**Comment by bashonly** [comment]
In the future, please avoid force-pushing in the middle of review. You can just do a merge commit

It will be reviewed when time permits

File: README.md [hunk 1]
```diff
@@ -1214,6 +1217,14 @@ To activate authentication with the `.netrc` file you should pass `--netrc` to y
 
 The default location of the .netrc file is `~` (see below).
 
+### Authentication with a custom shell command
+
+As an alternative to using the `.netrc` file, which has the disadvantage of keeping your passwords in a plain text file, you can configure a custom shell command to provide the credentials for an extractor. This is done by providing the `--netrc-cmd` parameter, it shall output the credentials in the netrc format and return `0` on success, other values will be treated as an error. `{}` in the command will be replaced by the name of the extractor to make it possible to select the credentials for the right extractor.
+E.g. for using the `pass` password manager:
+```
+yt-dlp --netrc-cmd "pass ytldp/youtube" https://www.youtube.com/watch?v=BaW_jenozKc
```
Comment (pukkandan) [inline README.md:1230]: This is wrong, no?

**Comment by pukkandan** [comment]
https://github.com/yt-dlp/yt-dlp/pull/6682/commits/48525a63ae3a67133f0218b90ba11d72dedef4ea is not fully tested coz Ive gtg. Once it is, and the doc is corrected, this can be merged.

**Comment by NDagestad** [comment]
From my testing this works without problem. Maybe the example is not the most pertinent though, maybe replacing it with `--netrc-cmd "gpg --decrypt ~/.authinfo.gpg 2> /dev/null"` like proposed in #1706 would make sens to more people.

**Comment by pukkandan** [comment]
> Maybe the example is not the most pertinent though, maybe replacing it with `--netrc-cmd "gpg --decrypt ~/.authinfo.gpg 2> /dev/null"` like proposed in #1706 would make sens to more people.

I agree that is easier to understand.

**Pull request merged**